client/question/services/FeedbackGeneratorService.js

```diff
@@ -24,18 +24,26 @@ tie.factory('FeedbackGeneratorService', [
   'RUNTIME_ERROR_FEEDBACK_MESSAGES', 'WRONG_LANGUAGE_ERRORS', 'LANGUAGE_PYTHON',
   'CLASS_NAME_AUXILIARY_CODE', 'CLASS_NAME_SYSTEM_CODE', 'PARAGRAPH_TYPE_TEXT',
   'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
-  'PYTHON_PRIMER_BUTTON_NAME', 'ERROR_COUNTER_LANGUAGE_UNFAMILIARITY',
-  'ERROR_COUNTER_SAME_RUNTIME', 'UNFAMILIARITY_THRESHOLD',
-  'FEEDBACK_CATEGORIES',
+  'PARAGRAPH_TYPE_OUTPUT', 'PYTHON_PRIMER_BUTTON_NAME',
+  'ERROR_COUNTER_LANGUAGE_UNFAMILIARITY', 'ERROR_COUNTER_SAME_RUNTIME',
+  'UNFAMILIARITY_THRESHOLD', 'FEEDBACK_CATEGORIES',
   function(
     FeedbackObjectFactory, TranscriptService, ReinforcementGeneratorService,
     CODE_EXECUTION_TIMEOUT_SECONDS, SUPPORTED_PYTHON_LIBS,
     RUNTIME_ERROR_FEEDBACK_MESSAGES, WRONG_LANGUAGE_ERRORS, LANGUAGE_PYTHON,
     CLASS_NAME_AUXILIARY_CODE, CLASS_NAME_SYSTEM_CODE, PARAGRAPH_TYPE_TEXT,
     PARAGRAPH_TYPE_CODE, PARAGRAPH_TYPE_SYNTAX_ERROR,
-    PYTHON_PRIMER_BUTTON_NAME, ERROR_COUNTER_LANGUAGE_UNFAMILIARITY,
-    ERROR_COUNTER_SAME_RUNTIME, UNFAMILIARITY_THRESHOLD,
-    FEEDBACK_CATEGORIES) {
+    PARAGRAPH_TYPE_OUTPUT, PYTHON_PRIMER_BUTTON_NAME,
+    ERROR_COUNTER_LANGUAGE_UNFAMILIARITY, ERROR_COUNTER_SAME_RUNTIME,
+    UNFAMILIARITY_THRESHOLD, FEEDBACK_CATEGORIES) {
+
+    /**
+     * Object used to keep track of how many corrective feedback (values)
+     * were generated for a given test suite (keys).
+     *
+     * @type {Object.<string, number>}
+     */
+    var attemptCounter = {};
```

Comment:
Ditto; the name needs to be more specific. Perhaps suiteIdToNumAttempts or similar. That said -- this is not the same as the total number of attempts made by the learner, right? It only counts those attempts which "trickle down" to the correctness-test level -- and that needs to be made clearer in the name + docstring.

Thinking about it more, I'd recommend a change in framing here: instead of mapping suite ID to number of attempts, instead map suite ID to one of three named string constants denoting the stages of feedback (CORRECTNESS_TEST_INPUT_GIVEN, CORRECTNESS_TEST_EXPECTED_OUTPUT_GIVEN, CORRECTNESS_TEST_OBSERVED_OUTPUT_GIVEN). This is so we just store the minimal "status" information necessary to give the appropriate feedback and determine the next status to use.

Reply:
How about correctnessFeedbackCounterBySuiteId?

I'm not sure I follow your suggestion. This appears to be more complex and granular than for the purpose I was using this.

Reply:
Will update per discussion.